Introduce naming conventions

PaleoCrafter opened the pull request:
## Motivation
With the increasing popularity of mapping things and the overall influx of changes, keeping certain quality standards becomes harder. There's been a lot of cleanup done in recent versions, but we still get bad *new* names, as can be seen from various issues created about bad 1.13 names.

Another goal we should strive for is consistency. Names are ultimately subjective, but a crowd sourced effort ought to adhere to a common style.

One of the issues people new to the mapping process face (which also effects long-time contributors) is the lack of a central place documenting conventions. There's tidbits spread out across issue comments and (private) conversations, but we do not have one definite source of truth.

## Approach
In order to tackle this, I've started working on a document outlining established naming conventions I've identified in issues and existing names. Some of them were taken from historical developments, things like `Obj` suffixes mostly are a thing of the past. They're still listed to be as clear as possible.

Do note that these are supposed to be **conventions**, not **rules**. There always will be exceptions, we can't provide more than a general guideline.

## Progress
The basics should be done, but I'd like to add some more examples. Suggestions are welcome, I'm not creative enough to come up with anything good.

All conventions are of course up for discussion. If you've got any input on formatting, structure or wording, please let me know.

bs2609 commented:
Can this be emphasised *even more*?
`If you do not know what to name something then *do not name it*. Let someone else name it.`


Other points under "things to consider before submitting a name":
* `please look at the thing in context, check the usages as well as the definition`
* `check the defining class of a method`

Mainly that first note, however, as it relates to the influx of bad **new** names.

Perhaps a "before you start"-type section before "general conventions"?

Runemoro commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+     - `has`
+     - `is`
+     - `should`
+     - `was`
```
Rather than listing 6 possible prefixes, why not make this more general and accept any third person verb as a prefix? (`contains`, `will`, `cannot`, `provides`, etc.).

Also, some methods (such as `Set.add`) do an action and return a boolean depending on whether or not the action was successful. The third-person prefixes should only be used for methods that return a boolean and can't modify the state of the object.

I suggest changing this to:

> Methods that return a `boolean` and are pure (don't affect the object or parameters) should always start with a verb in the third person (such as `can`, `does`, `has` or `is`).

kashike replied:
The 6 prefixes are there to list the common ones. I'm fine with your suggestion, but I would still like to list these ones as they are the most common.

PaleoCrafter replied:
I think a solution to this would be to say that `boolean` getters (I think that's a good compromise, they don't necessarily have to be side-effect-less, much less pure, as purity would imply always getting the same result). For determining what prefix to choose, I think "whatever would be used in a conditional (if) clause" makes sense.

Runemoro commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+ - 'Event handling' methods should follow an `on<Noun>` theme where the noun is the event's name
+   (with redundancies removed).
+ - Methods which read or write NBT should be prefixed `read` and `write` respectively.
+ - Methods which deseralize or serialize JSON should have a `deserialize` or `serialize` prefix accordingly. 
```
Maybe clarify the difference between `read`/`write` and `deserialize`/`serialize`:
The "read"/"write" prefixes should be used when no new instances are created:
 - `something.read(NBTTagCompound tag)`
 - `something.write(PacketBuf buffer)`

The "serialize"/"deserialize" prefixes should be used when a new instance (of either the serialized or deserialized object) is created:
 - `Something something = somethingSerializer.deserialize(json)`
 - `JsonObject json = something.serializeToJson()`
 - `JsonObject json = somethingSerializer.serialize(something)`

This seems to be the convention we're using and is also consistent with how java and gson use the word "deserialize".

kashike replied:
`read` and `write` should be used for NBT and network methods, while `serialize` and `deserialize` should be used for JSON methods.

Runemoro replied:
> read and write should be used for NBT and network methods, while serialize and deserialize should be used for JSON methods.

Json uses the words `serialize`/`deserialize` because unlike the NBT and network methods, it doesn't just load data from a buffer or NBT tag into an existing instance, but instead creates a new instance of the object from the json.

read = load data into an *existing* object
write = save the object to a buffer or NBT tag *passed as a parameter*
deserialize = *create a new object* from a serialized form (json in this case)
serialize = *create* a serialized form of the object

This works with the definition of [serialization](https://en.wikipedia.org/wiki/Serialization), which is the conversion between serialized and deserialized objects, not just loading data into an object from a tag/stream.

Runemoro commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+
+### Examples
+ - The method for handling block right clicks is called `onActivation` (noun form, no redundancy),
+   not `onBlockActivated` (past participle, redundant `Block` infix).
```
What about adding some rules about redundancy:
 - The containing class's name (and preceding prepositions) is omited from a method's name (`Container.addToContainer` -> `Container.add`)
 - The first parameter's type's name (but not preceding preposition) is omitted from a method's name (`something.addToSet(Set<?> set)` -> `something.addTo(Set<?> set)`, but not `something.add(Set<?> set)` since that has a different meaning of adding `set` to `something`)

PaleoCrafter replied:
I've added "avoiding redundancies" as a point under "General".

pau101 commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+
+General
+-------
+ - All names use US spelling.
```
"US spelling" is a bit weird, the more common way of phrasing that is "American English spelling" .

pau101 commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+Getting Started
+---------------
+ - **If you do not know what to name something then *do not name it*. Let someone else do so.**
+ - Names should reflect intent, *not* implementation. Check usages rather than just looking at the definition.
```
It would also be good to mention when naming methods to not only check usages, but also still consider implementation, especially when multiple types are providing implementations. This helps to further reveal intended semantics.

Pokechu22 commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+ - If a name clashes with that of a field or a type name, it should be suffixed with `In`
+ - Parameters of certain types should always get named according to the following list:
+     - `BlockPos` arguments should be named `pos` when there is a single one, and `*Pos` when there are multiple
+     - Parameters of the `IWorldReaderBase` and related type should be `worldIn`
```
I think it's important to qualify this with a "where the behavior is used simply as the world to perform the operation on" (and a similar thing for `BlockPos`).  For example, `p_72375_2_` for `PlayerList.preparePlayer` used to be named `worldIn`, but it's actually the world the player was in _previously_ and has been renamed to `oldWorld`.  Basically, look at it first and make sure the usage is conventional; this isn't an "always" rule.

gabizou replied:
Agreed, this shouldn't be an always rule, and there's cases where the world is not a world, but a `WorldGenRegion`, which I'd say is more applicable to `region`, but that's an argument outside the scope of this PR.

PaleoCrafter replied:
I've added this as further clarification.

Pokechu22 commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+   do not include it in the name. This especially applies to fields and methods in a class, where you do not
+   need to have the defining class's name in the member's name.
+ - Prefer clarification in the documentation comment over verbose names. Conversely, do not give fields,
+   methods, or method parameters useless comments.
```
Defining what "useless" is is somewhat important.  Describing what it does is (IMO) always useful even if redundant, except in perhaps the most obvious of cases.  On the other hand, a description of mechanics for an individual thing is generally not useful (e.g. for fields in `Blocks` or `Enchantments`), although I'd say it still is useful for mechanics that are fairly technical in nature/examples of where something might be used.

PaleoCrafter replied:
I've switched to `needless` and added a few exemplary details that should go into the comment.

Pokechu22 commented:
Actually, I think your statement from the post body is useful here:

> Do note that these are supposed to be **conventions**, not **rules**. There always will be exceptions, we can't provide more than a general guideline.

To me it feels like the actual document is written as rules and not conventions, for instance with use of "always".  Additional hedges (such as "should generally" at least for the parameters case and the boolean method one) would make it a bit clearer.

gabizou commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+ - Enums should be named directly after the things they are enumerating, in singular form.
+   There is no `Enum` prefix/suffix.
+
+**Note:** Long-established names where prefixes are used (subclasses `Block`, `Item` etc.)
```
Perhaps you should mention which long established names should be referred to, in the example below, you suffix the bookshelf  with a `Container` of type `ShelfContainer`.

PaleoCrafter replied:
Removed for now with the apparent support for suffixes. If that issue gets decided differently, I will update accordingly.

gabizou commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+
+Method Parameter Names
+----------------------
+ - If a name clashes with that of a field or a type name, it should be suffixed with `In`
```
This is somewhat conflicting because up to a point, you can name method parameters without the `in` prefix just fine, and then name a field the same name as method parameters without issue. Only after you name a field with the same name as you're intending to name a method parameter (like, say, `setPos(BlockPos pos) { this.pos = pos; }`) would you be unable to set the method parameter name without some prefix/suffix (in this case, the guide is to suffix).

Pokechu22 replied:
Do the field conflict restrictions even still make sense?  I'm pretty sure fernflower always adds `this`...

gabizou replied:
They don't make sense these days, but I guess for some people it's more legible.

Pokechu22 replied:
I've created #770 for discussing whether or not that check should be removed/is still needed.

PaleoCrafter replied:
Thanks for that. I'll leave that item in for now and we'll see what comes of your ticket.

gabizou commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+----------------------
+ - If a name clashes with that of a field or a type name, it should be suffixed with `In`
+ - Parameters of certain types should always get named according to the following list:
+     - `BlockPos` arguments should be named `pos` when there is a single one, and `*Pos` when there are multiple
```
How does `*Pos` work? `aPos`, `bPos`, `originalPos`, `additionalPos`?

PaleoCrafter replied:
Yep, I'll clarify that.

PaleoCrafter commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,75 @@
+
+### Examples
+ - For a new shelf block with an inventory, appropriate names would be `BlockShelf`, `TileEntityShelf` and
+   `ShelfContainer`.
```
I'm really not satisfied with the example here, myself. The mix of prefixing and suffixing really doesn't help it. @kashike, do we have a list of all "these should stay prefixes" names?

PaleoCrafter replied:
Since suffixes are winning right now, I've changed this up again.

Pokechu22 commented on CONVENTIONS.md:
```diff
@@ -0,0 +1,81 @@
+ - Acronyms should generally be written all-lowercase, while normal camel case rules still apply.
+   For instance, "identifier" becomes `id` as standalone word or at the beginning of a camel case phrase and
+   `Id` when used within a phrase.<br>
+   There are various exceptions to this rule which apply only when the word does not start with them:
```
What about UUID?  Currently it's rather inconsistent with some uses being `UUID`, some being `Uuid`, and some not compacting it at all.  Mojang seems to use `UUID` for `Realms.getUUID` (but there's also `Realms.uuidToName`).

kashike replied:
`UniqueId` is what we tend to use.

PaleoCrafter replied:
I've added an item about `UUID` -> `UniqueId`

PaleoCrafter commented:
I've updated the document with RFC-style terminology. Furthermore, I've added some rules with regards to casing, as that should also be consistent.

With the way the voting issues currently go, I've also modified the sections that are affected. If the outcome on Monday is any different, I will reflect this here.

LexManos commented:
All looks good to me and the votes are over, So going to pull this.
